imi/src/main/java/org/motechproject/nms/imi/service/impl/CsrValidatorServiceImpl.java

```diff
@@ -0,0 +1,181 @@
+import org.motechproject.nms.kilkari.domain.SubscriptionPackMessage;
+import org.motechproject.nms.kilkari.dto.CallSummaryRecordDto;
+import org.motechproject.nms.kilkari.service.SubscriptionService;
+import org.motechproject.nms.region.domain.Circle;
+import org.motechproject.nms.region.domain.LanguageLocation;
+import org.motechproject.nms.region.service.CircleService;
+import org.motechproject.nms.region.service.LanguageLocationService;
+import org.motechproject.server.config.SettingsFacade;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@Service("csrValidatorService")
+public class CsrValidatorServiceImpl implements CsrValidatorService {
+
+    private static final String MAX_CDR_ERROR_COUNT = "imi.max_cdr_error_count";
+    private static final int MAX_CDR_ERROR_COUNT_DEFAULT = 100;
+
+    private Map<String, Set<String>> weekIds;
+    private Map<String, Set<String>> contentFileNames;
+    private Set<String> languageLocationCodes;
+    private Set<String> circles;
+
+    private SettingsFacade settingsFacade;
+    private SubscriptionService subscriptionService;
+    private LanguageLocationService languageLocationService;
+    private CircleService circleService;
+
+
+    @Autowired
+    public CsrValidatorServiceImpl(@Qualifier("imiSettings") SettingsFacade settingsFacade,
+                                   SubscriptionService subscriptionService,
+                                   LanguageLocationService languageLocationService,
+                                   CircleService circleService) {
+        this.settingsFacade = settingsFacade;
+        this.subscriptionService = subscriptionService;
+        this.languageLocationService = languageLocationService;
+        this.circleService = circleService;
+    }
+
+
+    private void loadPacks() {
+        weekIds = new HashMap<>();
+        contentFileNames = new HashMap<>();
+        for (SubscriptionPack pack : subscriptionService.getSubscriptionPacks()) {
+            Set<String> packWeekIds = new HashSet<String>();
+            Set<String> packContentFileNames = new HashSet<String>();
+            for (SubscriptionPackMessage message :pack.getMessages()) {
+                packWeekIds.add(message.getWeekId());
+                packContentFileNames.add(message.getMessageFileName());
+            }
+            weekIds.put(pack.getName(), packWeekIds);
+            contentFileNames.put(pack.getName(), packContentFileNames);
+        }
+    }
+
+
+    private void loadLanguageLocationCodes() {
+        languageLocationCodes = new HashSet<String>();
+        for (LanguageLocation languageLocation : languageLocationService.getAll()) {
```

Comment:
Since you just want the names, I wonder if we will need to create a method that returns just the language location codes.  My concern is since all of our related objects are in the defaultFetchGroup that to get all the codes we'll also have to load all the related objects (and maybe their related objects)  I have the same concerns about Circles.  This might be a Sprint 4 issue to test.